File: app/src/main/java/com/waray/spendhound/DebtTransactionAdapter.java
package com.waray.spendhound;

import android.annotation.SuppressLint;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.card.MaterialCardView;

import java.util.ArrayList;

public class DebtTransactionAdapter extends RecyclerView.Adapter<DebtTransactionAdapter.ViewHolder> {
    private final ArrayList<BorrowTransaction> borrowTransactionList;
    private ArrayList<Integer> checkedPositions;
    private OnItemClickListener clickListener;
    private OnBorrowerActionListener borrowerActionListener;

    public interface OnItemClickListener {
        void onItemClick(BorrowTransaction transaction, int position);
    }

    public interface OnBorrowerActionListener {
        void onPayClicked(BorrowTransaction transaction, int position);
        void onRemoveClicked(BorrowTransaction transaction, int position);
        void onTryAgainClicked(BorrowTransaction transaction, int position);
    }

    public DebtTransactionAdapter(ArrayList<BorrowTransaction> borrowTransactionList) {
        this.borrowTransactionList = borrowTransactionList;
        checkedPositions = new ArrayList<>();
    }

    public DebtTransactionAdapter(ArrayList<BorrowTransaction> borrowTransactionList, OnItemClickListener clickListener) {
        this.borrowTransactionList = borrowTransactionList;
        this.clickListener = clickListener;
        checkedPositions = new ArrayList<>();
    }

    public DebtTransactionAdapter(ArrayList<BorrowTransaction> borrowTransactionList, OnBorrowerActionListener borrowerActionListener) {
        this.borrowTransactionList = borrowTransactionList;
        this.borrowerActionListener = borrowerActionListener;
        checkedPositions = new ArrayList<>();
    }

    public void setOnItemClickListener(OnItemClickListener listener) {
        this.clickListener = listener;
    }

    public void setOnBorrowerActionListener(OnBorrowerActionListener listener) {
        this.borrowerActionListener = listener;
    }

    // Add this method to retrieve a BorrowTransaction by its position
    public BorrowTransaction getBorrowTransaction(int position) {
        return borrowTransactionList.get(position);
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View itemView = LayoutInflater.from(parent.getContext()).inflate(R.layout.debt_row_layout, parent, false);
        return new ViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, @SuppressLint("RecyclerView") int position) {
        BorrowTransaction transaction = borrowTransactionList.get(position);

        // Bind data to the ViewHolder's views
        holder.debtDateTV.setText(transaction.getDate());
        holder.debtBorroweeTV.setText(transaction.getBorrowee());
        holder.debtAmountBorrowedTV.setText(transaction.getBorrowedAmountStr());
        holder.debtStatusTV.setText(transaction.getStatus());

        // Hide all action layouts by default
        holder.unpaidActionsLayout.setVisibility(View.GONE);
        holder.declinedActionsLayout.setVisibility(View.GONE);
        holder.paymentSentDateTV.setVisibility(View.GONE);

        // Set status color based on status value
        String status = transaction.getStatus();
        int statusColor;
        boolean isPendingStatus = false;
        boolean isPendingPayment = false;
        boolean isPaid = false;
        boolean isUnpaid = false;
        boolean isDeclined = false;
        if ("Paid".equalsIgnoreCase(status)) {
            statusColor = ContextCompat.getColor(holder.itemView.getContext(), R.color.green);
            isPaid = true;
        } else if ("Pending".equalsIgnoreCase(status)) {
            statusColor = ContextCompat.getColor(holder.itemView.getContext(), R.color.yellow);
        } else if ("Paid Partially".equalsIgnoreCase(status)) {
            statusColor = ContextCompat.getColor(holder.itemView.getContext(), R.color.yellow);
        } else if ("Pending Payment".equalsIgnoreCase(status)) {
            statusColor = ContextCompat.getColor(holder.itemView.getContext(), R.color.yellow);
            isPendingPayment = true;
        } else if ("For Lender Approval".equalsIgnoreCase(status)) {
            statusColor = ContextCompat.getColor(holder.itemView.getContext(), R.color.blue);
            isPendingStatus = true;
        } else if ("Declined".equalsIgnoreCase(status)) {
            statusColor = ContextCompat.getColor(holder.itemView.getContext(), R.color.red);
            isDeclined = true;
        } else if ("Unpaid".equalsIgnoreCase(status)) {
            statusColor = ContextCompat.getColor(holder.itemView.getContext(), R.color.red);
            isUnpaid = true;
        } else {
            statusColor = ContextCompat.getColor(holder.itemView.getContext(), R.color.red);
        }
        holder.debtStatusTV.setTextColor(statusColor);

        // Show action buttons based on status
        if (isUnpaid) {
            holder.unpaidActionsLayout.setVisibility(View.VISIBLE);
        } else if (isDeclined) {
            holder.declinedActionsLayout.setVisibility(View.VISIBLE);
        }

        // Show payment sent date for Paid status
        if (isPaid && transaction.getPaymentSentDate() != null && !transaction.getPaymentSentDate().isEmpty()) {
            holder.paymentSentDateTV.setVisibility(View.VISIBLE);
            holder.paymentSentDateTV.setText("Sent: " + transaction.getPaymentSentDate());
        }

        // Cast to MaterialCardView for elevation and background
        MaterialCardView cardView = (MaterialCardView) holder.itemView;
        float density = holder.itemView.getContext().getResources().getDisplayMetrics().density;

        // Set background tint and elevation based on status
        if (isPaid) {
            cardView.setCardBackgroundColor(ContextCompat.getColor(holder.itemView.getContext(), R.color.paid_bg));
            cardView.setCardElevation(4 * density);
        } else if (isUnpaid) {
            cardView.setCardBackgroundColor(ContextCompat.getColor(holder.itemView.getContext(), R.color.unpaid_bg));
            cardView.setCardElevation(4 * density);
        } else if (isPendingPayment) {
            cardView.setCardBackgroundColor(ContextCompat.getColor(holder.itemView.getContext(), R.color.pending_payment_bg));
            cardView.setCardElevation(4 * density);
        } else if (isPendingStatus) {
            cardView.setCardBackgroundColor(ContextCompat.getColor(holder.itemView.getContext(), R.color.pending_approval_bg));
            cardView.setCardElevation(4 * density);
        } else if (isDeclined) {
            cardView.setCardBackgroundColor(ContextCompat.getColor(holder.itemView.getContext(), R.color.unpaid_bg));
            cardView.setCardElevation(4 * density);
        } else {
            cardView.setCardBackgroundColor(ContextCompat.getColor(holder.itemView.getContext(), R.color.whitest));
            cardView.setCardElevation(4 * density);
        }

        // Set click listeners for action buttons
        holder.payBtn.setOnClickListener(v -> {
            if (borrowerActionListener != null) {
                borrowerActionListener.onPayClicked(transaction, holder.getAdapterPosition());
            }
        });

        holder.removeBtn.setOnClickListener(v -> {
            if (borrowerActionListener != null) {
                borrowerActionListener.onRemoveClicked(transaction, holder.getAdapterPosition());
            }
        });

        holder.tryAgainBtn.setOnClickListener(v -> {
            if (borrowerActionListener != null) {
                borrowerActionListener.onTryAgainClicked(transaction, holder.getAdapterPosition());
            }
        });

        // Set click listener for pending items to navigate to PendingStatusActivity
        final boolean finalIsPendingStatus = isPendingStatus;
        holder.itemView.setOnClickListener(v -> {
            if (clickListener != null && finalIsPendingStatus) {
                clickListener.onItemClick(transaction, position);
            }
        });
    }


    @Override
    public int getItemCount() {
        return borrowTransactionList.size();
    }

    public ArrayList<Integer> getCheckedPositions() {
        return checkedPositions;
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        public TextView debtDateTV;
        public TextView debtBorroweeTV;
        public TextView debtAmountBorrowedTV;
        public TextView debtStatusTV;
        public TextView paymentSentDateTV;
        public LinearLayout unpaidActionsLayout;
        public LinearLayout declinedActionsLayout;
        public TextView payBtn;
        public TextView removeBtn;
        public TextView tryAgainBtn;

        public ViewHolder(View itemView) {
            super(itemView);
            debtDateTV = itemView.findViewById(R.id.debtDateTV);
            debtBorroweeTV = itemView.findViewById(R.id.debtBorroweeTV);
            debtAmountBorrowedTV = itemView.findViewById(R.id.debtAmountBorrowedTV);
            debtStatusTV = itemView.findViewById(R.id.debtStatusTV);
            paymentSentDateTV = itemView.findViewById(R.id.debtPaymentSentDateTV);
            unpaidActionsLayout = itemView.findViewById(R.id.unpaidActionsLayout);
            declinedActionsLayout = itemView.findViewById(R.id.declinedActionsLayout);
            payBtn = itemView.findViewById(R.id.payBtn);
            removeBtn = itemView.findViewById(R.id.removeBtn);
            tryAgainBtn = itemView.findViewById(R.id.tryAgainBtn);
        }
    }
}

File: app/src/main/java/com/waray/spendhound/DebtTransactionAdapter.kt
package com.waray.spendhound

import android.annotation.SuppressLint
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.card.MaterialCardView

/**
 * Adapter for displaying debt transactions where the current user is the borrower.
 */
class DebtTransactionAdapter(
    private val borrowTransactionList: List<BorrowTransaction?>,
    private var borrowerActionListener: OnBorrowerActionListener? = null
) : RecyclerView.Adapter<DebtTransactionAdapter.ViewHolder>() {

    private var clickListener: OnItemClickListener? = null

    interface OnItemClickListener {
        fun onItemClick(transaction: BorrowTransaction?, position: Int)
    }

    interface OnBorrowerActionListener {
        fun onPayClicked(transaction: BorrowTransaction?, position: Int)
        fun onRemoveClicked(transaction: BorrowTransaction?, position: Int)
        fun onTryAgainClicked(transaction: BorrowTransaction?, position: Int)
    }

    /**
     * Secondary constructor for setting the item click listener.
     */
    constructor(
        borrowTransactionList: List<BorrowTransaction?>,
        clickListener: OnItemClickListener?
    ) : this(borrowTransactionList) {
        this.clickListener = clickListener
    }

    fun setOnItemClickListener(listener: OnItemClickListener?) {
        this.clickListener = listener
    }

    fun setOnBorrowerActionListener(listener: OnBorrowerActionListener?) {
        this.borrowerActionListener = listener
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val itemView = LayoutInflater.from(parent.context)
            .inflate(R.layout.debt_row_layout, parent, false)
        return ViewHolder(itemView)
    }

    override fun onBindViewHolder(holder: ViewHolder, @SuppressLint("RecyclerView") position: Int) {
        val transaction = borrowTransactionList[position] ?: return

        // Bind data using property access
        holder.debtDateTV.text = formatDate(transaction.date)
        holder.debtBorroweeTV.text = transaction.borroweeDisplayName ?: transaction.borrowee
        holder.debtAmountBorrowedTV.text = CurrencyUtils.formatAmountWithCurrency(transaction.borrowedAmountStr ?: "0")
        holder.debtStatusTV.text = transaction.status

        // Reset visibility for layouts
        holder.unpaidActionsLayout.visibility = View.GONE
        holder.declinedActionsLayout.visibility = View.GONE
        holder.paymentSentDateTV.visibility = View.GONE

        val status = transaction.status
        var isPendingStatus = false
        var isPendingPayment = false
        var isPaid = false
        var isUnpaid = false
        var isDeclined = false

        val statusColor = when {
            "Paid".equals(status, ignoreCase = true) -> {
                isPaid = true
                ContextCompat.getColor(holder.itemView.context, R.color.green)
            }
            "Pending".equals(status, ignoreCase = true) || 
            "Paid Partially".equals(status, ignoreCase = true) ||
            "Pending Payment".equals(status, ignoreCase = true) -> {
                if ("Pending Payment".equals(status, ignoreCase = true)) isPendingPayment = true
                ContextCompat.getColor(holder.itemView.context, R.color.yellow)
            }
            "For Lender Approval".equals(status, ignoreCase = true) -> {
                isPendingStatus = true
                ContextCompat.getColor(holder.itemView.context, R.color.blue)
            }
            "Declined".equals(status, ignoreCase = true) -> {
                isDeclined = true
                ContextCompat.getColor(holder.itemView.context, R.color.red)
            }
            "Unpaid".equals(status, ignoreCase = true) -> {
                isUnpaid = true
                ContextCompat.getColor(holder.itemView.context, R.color.red)
            }
            else -> ContextCompat.getColor(holder.itemView.context, R.color.red)
        }
        
        holder.debtStatusTV.setTextColor(statusColor)

        // Conditional UI visibility
        if (isUnpaid) {
            holder.unpaidActionsLayout.visibility = View.VISIBLE
        } else if (isDeclined) {
            holder.declinedActionsLayout.visibility = View.VISIBLE
        }

        if (isPaid && !transaction.paymentSentDate.isNullOrEmpty()) {
            holder.paymentSentDateTV.visibility = View.VISIBLE
            holder.paymentSentDateTV.text = holder.itemView.context.getString(R.string.payment_sent_format, transaction.paymentSentDate)
        }

        // Card UI management
        val cardView = holder.itemView as MaterialCardView
        val density = holder.itemView.context.resources.displayMetrics.density
        
        val bgColorRes = when {
            isPaid -> R.color.paid_bg
            isUnpaid || isDeclined -> R.color.unpaid_bg
            isPendingPayment -> R.color.pending_payment_bg
            isPendingStatus -> R.color.pending_approval_bg
            else -> R.color.whitest
        }
        
        cardView.setCardBackgroundColor(ContextCompat.getColor(holder.itemView.context, bgColorRes))
        cardView.cardElevation = 4 * density

        // Action Listeners
        holder.payBtn.setOnClickListener {
            borrowerActionListener?.onPayClicked(transaction, holder.bindingAdapterPosition)
        }

        holder.removeBtn.setOnClickListener {
            borrowerActionListener?.onRemoveClicked(transaction, holder.bindingAdapterPosition)
        }

        holder.tryAgainBtn.setOnClickListener {
            borrowerActionListener?.onTryAgainClicked(transaction, holder.bindingAdapterPosition)
        }

        holder.itemView.setOnClickListener {
            if (isPendingStatus) {
                clickListener?.onItemClick(transaction, holder.bindingAdapterPosition)
            }
        }
    }

    override fun getItemCount(): Int = borrowTransactionList.size

    companion object {
        fun formatDate(raw: String?): String {
            if (raw.isNullOrBlank()) return ""
            val formats = listOf(
                "yyyy-MM-dd'T'HH:mm:ss.SSSXXX",
                "yyyy-MM-dd'T'HH:mm:ssXXX",
                "yyyy-MM-dd'T'HH:mm:ss",
                "MMMM-dd-yyyy"
            )
            for (fmt in formats) {
                try {
                    val date = java.text.SimpleDateFormat(fmt, java.util.Locale.getDefault()).parse(raw)
                    if (date != null) return java.text.SimpleDateFormat("MMM dd, yyyy", java.util.Locale.getDefault()).format(date)
                } catch (_: Exception) {}
            }
            return raw
        }
    }

    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val debtDateTV: TextView = itemView.findViewById(R.id.debtDateTV)
        val debtBorroweeTV: TextView = itemView.findViewById(R.id.debtBorroweeTV)
        val debtAmountBorrowedTV: TextView = itemView.findViewById(R.id.debtAmountBorrowedTV)
        val debtStatusTV: TextView = itemView.findViewById(R.id.debtStatusTV)
        val paymentSentDateTV: TextView = itemView.findViewById(R.id.debtPaymentSentDateTV)
        val unpaidActionsLayout: LinearLayout = itemView.findViewById(R.id.unpaidActionsLayout)
        val declinedActionsLayout: LinearLayout = itemView.findViewById(R.id.declinedActionsLayout)
        val payBtn: TextView = itemView.findViewById(R.id.payBtn)
        val removeBtn: TextView = itemView.findViewById(R.id.removeBtn)
        val tryAgainBtn: TextView = itemView.findViewById(R.id.tryAgainBtn)
    }
}
